File: java/leetcode/AbsolutePath.java
//Leetcode problem 388 Longest Absolute File Path
//Solution written by Xuqiang Fang on 15 June, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class Solution{
    public int lengthLongestPath(String input) {
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        int max = 0;
        for(String s : input.split("\n")){
            //System.out.println(s);
            int lev = s.lastIndexOf("\t")+1;
            //System.out.println(lev);
            while(lev+1 < stack.size()){
                stack.pop();
            }
            int len = stack.peek() + s.length() -lev + 1;
            stack.push(len);
            if(s.contains(".")){
                max = Math.max(max, len-1);
            }
        } 
        return max;
    }
}

public class AbsolutePath{
	public static void main(String[] args){
		Solution s = new Solution();
        String input = "dir\n\tsubdir1\n\t\tfile1.ext\n\t\tsubsubdir1\n\tsubdir2\n\t\tsubsubdir2\n\t\t\tfile2.ext";
        System.out.println(s.lengthLongestPath(input));
        String input2 = "dir\n\tsubdir1\n\tsubdir2\n\t\tfile.ext";
        System.out.println(s.lengthLongestPath(input2));
	}
}

File: cpp/leetcode/AbsolutePath.cpp
//Leetcode Problem No 388 Longest Absolute File Path
//Solution written by Xuqiang Fang on 21 June, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    int lengthLongestPath(string input){
        int ans = 0;
        stack<int> st;
        st.push(0);
        for(auto& s : split(input, '\n')){
            int lev = s.find_last_of('\t') + 1;
            while(lev+1 < st.size()){
                st.pop();
            }
            int len = st.top() + s.length() - lev + 1;
            st.push(len);
            if (s.find('.') != -1){
                ans = max(len-1, ans);
            }
        }
        return ans;
    }
private:
    vector<string> split(string s, char c){
        vector<string> ans;
        int p = s.find(c);
        while(p != -1){
            ans.push_back(s.substr(0,p));
            s = s.substr(p+1);
            p = s.find(c);
        }
        ans.push_back(s);
        return ans;
    }
};

int main(){
    Solution s;
    string input{"dir\n\tsubdir1\n\t\tfile1.ext\n\t\tsubsubdir1\n\tsubdir2\n\t\tsubsubdir2\n\t\t\tfile2.ext"};
    cout << s.lengthLongestPath(input) << endl;
    cout << input << endl;
    return 0;
}
